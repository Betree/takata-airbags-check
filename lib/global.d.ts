import { messages } from "./i18n/messages";

declare global {
  namespace FormatjsIntl {
    interface Message {
      ids: keyof (typeof messages)["fr"];
    }
  }
}

declare module "*.json" {
  const value: any;
  export default value;
}
