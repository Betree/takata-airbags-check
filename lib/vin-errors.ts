// Mapped to the errors from https://github.com/mobito-tech-io/universal-vin-decoder/blob/main/src/vin.ts
const I18N_VIN_MESSAGES: Record<string, string> = {
  "VIN must be a string": "Format de VIN invalide",
  "VIN must be 17 characters long": "Le VIN doit contenir 17 caractÃ¨res",
  "VIN contain only letters & numbers except from I, O and Q":
    "Le VIN ne peut contenir que des chiffres et des lettres (sauf I, O et Q).",
};

export const translateVinError = (message: string): string => {
  return I18N_VIN_MESSAGES[message] || message;
};
