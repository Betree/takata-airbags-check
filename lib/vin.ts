import { decodeVIN, splitVIN, validateVIN } from "universal-vin-decoder";
import { words } from "lodash";
import {
  ManufacturersInfo,
  AffectedModel,
  AllManufacturers,
  Manufacturer,
  ManufacturerData,
  NeverAffectedManufacturers,
} from "./data/affected-cars";

type CarReference = {
  ma: string; // model abbreviation
  me?: string; // model extended name
};

type CarReferences = {
  [key: string]: CarReference;
};

import rawCarReferences from "./data/car-references.json";
const carReferences = rawCarReferences as CarReferences;

export const validateVINI18n = (
  vin: string
):
  | { isValid: true }
  | {
      isValid: false;
      error: string;
    } => {
  const result = validateVIN(vin);
  if (result.isValid) {
    return { isValid: true };
  } else {
    return { isValid: false, error: result.error || "Unknown error" };
  }
};

const removeDiacritics = (
  manufacturer: Exclude<
    ReturnType<typeof decodeVIN>["info"],
    undefined
  >["manufacturer"]
): string => {
  return manufacturer.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
};

export const mapLibManufacturer = (
  manufacturer: string
): Manufacturer | undefined => {
  const cleanManufacturer = removeDiacritics(
    manufacturer.replace(/\s+/g, " ").trim()
  ).toLowerCase();
  const splitManufacturer = words(cleanManufacturer);
  for (const [manufacturer, settings] of Object.entries(ManufacturersInfo)) {
    if (
      settings.words?.some((word) => splitManufacturer.includes(word)) ||
      // For simple names, we check with word to make sure "Ford" doesn't match "Bedford"
      splitManufacturer.includes(manufacturer.toLowerCase()) ||
      // When the name includes a space (e.g. "Land Rover"), we permit a direct check
      (cleanManufacturer.includes(" ") &&
        manufacturer.includes(cleanManufacturer))
    ) {
      return manufacturer as Manufacturer;
    }
  }

  return undefined;
};

export type ModelInfo = {
  manufacturer: string;
  name: string;
  extendedName?: string;
};

const transformModel = (
  model: Record<string, string>
): ModelInfo | undefined => {
  return !model
    ? undefined
    : {
        manufacturer: model.ma,
        name: model.mo,
        extendedName: model.me,
      };
};

export const formatModelInfoName = (modelInfo: ModelInfo) => {
  if (!modelInfo) {
    return "Inconnu";
  } else {
    return [modelInfo.name, modelInfo.extendedName].filter(Boolean).join(" - ");
  }
};

const getAffectedCarByModel = (
  manufacturer: Manufacturer,
  modelInfo: ModelInfo
) => {
  const lowerCaseModel = modelInfo.name.toLowerCase();
  const lowerCaseModelExtendedName = modelInfo.extendedName?.toLowerCase();
  return ManufacturersInfo[manufacturer]?.affected?.find(
    (affected) =>
      affected.model.toLowerCase() === lowerCaseModel ||
      affected.model.toLowerCase().includes(lowerCaseModel) ||
      (lowerCaseModelExtendedName &&
        affected.model.toLowerCase().includes(lowerCaseModelExtendedName))
  );
};

export const checkAffectedCar = (
  vin: string
): {
  splitVIN: ReturnType<typeof splitVIN>;
  decodedVIN: ReturnType<typeof decodeVIN>;
} & (
  | {
      isAffected: undefined;
      reason: "UNKNOWN_MANUFACTURER";
    }
  | {
      isAffected: undefined;
      reason: "UNKNOWN_MODEL";
      links: ManufacturerData["links"];
      siblingReports: AffectedModel[];
    }
  | {
      isAffected: undefined;
      reason: "UNKNOWN_YEAR";
      links: ManufacturerData["links"];
      modelReport: AffectedModel;
      siblingReports: AffectedModel[];
    }
  | {
      isAffected: false;
      reason: "MANUFACTURER_NEVER_AFFECTED";
      modelInfo?: ModelInfo;
    }
  | {
      isAffected: false;
      reason: "NO_REPORT";
      links: ManufacturerData["links"];
      modelInfo?: ModelInfo;
      siblingReports: AffectedModel[];
    }
  | {
      isAffected: false;
      reason: "YEAR_NOT_IN_RANGE";
      links: ManufacturerData["links"];
      modelInfo?: ModelInfo;
      modelReport?: AffectedModel;
      siblingReports: AffectedModel[];
    }
  | {
      isAffected: true;
      manufacturer: Manufacturer;
      links: ManufacturerData["links"];
      modelReport: AffectedModel;
      siblingReports: AffectedModel[];
    }
) => {
  const rawVinInfo = splitVIN(vin);
  const decodedVIN = decodeVIN(vin);
  const base = { splitVIN: rawVinInfo, decodedVIN };

  // Check manufacturer
  if (!decodedVIN.info?.manufacturer) {
    return { ...base, isAffected: undefined, reason: "UNKNOWN_MANUFACTURER" };
  }

  // VIN year codes are circular over a 30 years period. For the Takata case in France, only
  // cars from 1998 and 2019 are affected (see https://www.ecologie.gouv.fr/rappel-airbag-takata).
  const checkToYear = Math.min(new Date().getFullYear(), 2028);
  const checkFromYear = checkToYear - 30;

  decodedVIN.info.possibleModelYears =
    decodedVIN.info.possibleModelYears?.filter(
      (year) => year >= checkFromYear && year <= checkToYear
    );

  const manufacturer = mapLibManufacturer(decodedVIN.info.manufacturer);
  if (!manufacturer) {
    return { ...base, isAffected: undefined, reason: "UNKNOWN_MANUFACTURER" };
  }

  const modelInfo = transformModel(carReferences[rawVinInfo.vds]);
  if (NeverAffectedManufacturers.includes(manufacturer)) {
    return {
      ...base,
      isAffected: false,
      reason: "MANUFACTURER_NEVER_AFFECTED",
      modelInfo,
    };
  }

  // Check model
  let siblingReports = ManufacturersInfo[manufacturer]?.affected || [];
  if (!modelInfo) {
    return {
      ...base,
      isAffected: undefined,
      reason: "UNKNOWN_MODEL",
      links: ManufacturersInfo[manufacturer]?.links || {},
      siblingReports,
    };
  }

  // Check model report
  const modelReport = getAffectedCarByModel(manufacturer, modelInfo);
  if (!modelReport) {
    return {
      ...base,
      isAffected: false,
      reason: "NO_REPORT",
      links: ManufacturersInfo[manufacturer]?.links || {},
      modelInfo,
      siblingReports,
    };
  }

  siblingReports = siblingReports.filter((report) => report !== modelReport);

  if (!decodedVIN.info?.possibleModelYears?.length) {
    return {
      ...base,
      isAffected: undefined,
      reason: "UNKNOWN_YEAR",
      links: ManufacturersInfo[manufacturer]?.links || {},
      modelReport,
      siblingReports,
    };
  }

  for (const year of decodedVIN.info.possibleModelYears) {
    if (modelReport.yearFrom > year || modelReport.yearTo < year) {
      return {
        ...base,
        isAffected: false,
        reason: "YEAR_NOT_IN_RANGE",
        links: ManufacturersInfo[manufacturer]?.links || {},
        modelInfo,
        modelReport,
        siblingReports,
      };
    }
  }

  return {
    ...base,
    isAffected: true,
    manufacturer,
    links: ManufacturersInfo[manufacturer]?.links || {},
    modelReport,
    siblingReports,
  };
};
