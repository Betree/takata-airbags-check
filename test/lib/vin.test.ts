import { readFileSync } from "fs";
import { checkAffectedCar, mapLibManufacturer } from "../../lib/vin";
import { Manufacturer } from "@/lib/data/affected-cars";

const buildLibEuropeanManufacturers = () => {
  const libPath = "./node_modules/universal-vin-decoder/dist/index.js";
  const libFileContent = readFileSync(libPath).toString();
  const manufacturers: string[] = [];

  let isEUManufacturersSection = false;
  for (const line of libFileContent.split("\n")) {
    // Check start/end tokens
    if (line === `  manufacturers2["SA9"] = "Morgan Motor Company";`) {
      isEUManufacturersSection = true;
    } else if (
      line ===
      `  manufacturers2["1A4"] = "Chrysler brand  MPV/SUV 2006-2009 only";`
    ) {
      return manufacturers;
    }

    // Add manufacturers to the list
    if (isEUManufacturersSection) {
      manufacturers.push(
        line.split("=")[1].trim().replace(/;$/, "").replace(/["']/g, "")
      );
    }
  }

  throw new Error("Failed to find start/end tokens for EU manufacturers");
};

describe("VIN checker", () => {
  describe("invalid VINs", () => {
    it("should reject VIN that's too short", () => {
      const result = checkAffectedCar("ABC123");
      expect(result).toStrictEqual({
        decodedVIN: {
          message: "VIN must be 17 characters long",
          isValid: false,
          vin: "ABC123",
        },
        splitVIN: {
          modelYear: "",
          vds: "123",
          vis: "",
          wmi: "ABC",
        },
        isAffected: undefined,
        reason: "UNKNOWN_MANUFACTURER",
      });
    });

    it("should reject VIN with invalid characters", () => {
      const result = checkAffectedCar("VF7$XHMZ6JT00000");
      expect(result).toStrictEqual({
        decodedVIN: {
          message: "VIN must be 17 characters long",
          isValid: false,
          vin: "VF7$XHMZ6JT00000",
        },
        splitVIN: {
          modelYear: "J",
          vds: "$XHMZ6",
          vis: "JT00000",
          wmi: "VF7",
        },
        isAffected: undefined,
        reason: "UNKNOWN_MANUFACTURER",
      });
    });

    it("should reject empty VIN", () => {
      const result = checkAffectedCar("");
      expect(result).toStrictEqual({
        decodedVIN: {
          isValid: false,
          vin: "",
          message: "VIN must be 17 characters long",
        },
        splitVIN: {
          modelYear: "",
          vds: "",
          vis: "",
          wmi: "",
        },
        isAffected: undefined,
        reason: "UNKNOWN_MANUFACTURER",
      });
    });
  });

  describe("year-based safety checks", () => {
    it("should identify Citroen C3 as safe due to year", () => {
      const result = checkAffectedCar("VF7SXHMZ6JT000000");
      expect(result).toStrictEqual({
        decodedVIN: {
          info: {
            country: "France",
            manufacturer: "CitroÃ«n",
            modelYear: "2018",
            possibleModelYears: [2018],
            region: "Europe",
          },
          isValid: true,
          vin: "VF7SXHMZ6JT000000",
        },
        isAffected: false,
        links: {
          guadeloupe:
            "https://www.citroen.gp/entretenir/campagnes-de-rappel.html",
          guyane: "https://www.citroen.gf/services/campagnes-de-rappel/",
          hexagon: "https://www.citroen.fr/entretenir/campagnes-de-rappel.html",
          martinique: "https://www.citroen.mq/services/campagnes-de-rappel",
          reunion: "https://citroen-campagne-de-rappel.re/",
        },
        modelInfo: {
          extendedName: "C3 III",
          manufacturer: "CITROEN",
          name: "C3",
        },
        modelReport: {
          model: "C3 II",
          yearFrom: 2008,
          yearTo: 2017,
        },
        reason: "YEAR_NOT_IN_RANGE",
        siblingReports: [
          {
            model: "C4 II",
            yearFrom: 2010,
            yearTo: 2018,
          },
          {
            model: "DS3 I",
            yearFrom: 2008,
            yearTo: 2019,
          },
          {
            model: "DS4 I",
            yearFrom: 2010,
            yearTo: 2017,
          },
          {
            model: "DS5",
            yearFrom: 2010,
            yearTo: 2018,
          },
        ],
        splitVIN: {
          modelYear: "J",
          vds: "SXHMZ6",
          vis: "JT000000",
          wmi: "VF7",
        },
      });
    });

    it("should identify Honda Jazz as safe due to year", () => {
      const result = checkAffectedCar("JHMGK3770HX000000");
      expect(result).toStrictEqual({
        decodedVIN: {
          info: {
            country: "Japan",
            manufacturer: "Honda car",
            modelYear: "1987",
            possibleModelYears: [2017],
            region: "Asia",
          },
          isValid: true,
          vin: "JHMGK3770HX000000",
        },
        isAffected: false,
        links: {
          guadeloupe: "https://auto.honda.fr/cars/owners/airbag-recall.html",
          guyane: "https://auto.honda.fr/cars/owners/airbag-recall.html",
          hexagon: "https://auto.honda.fr/cars/owners/airbag-recall.html",
          martinique: "https://auto.honda.fr/cars/owners/airbag-recall.html",
        },
        modelInfo: {
          extendedName: undefined,
          manufacturer: "HONDA",
          name: "JAZZ",
        },
        modelReport: {
          model: "Jazz I, II et III",
          yearFrom: 2000,
          yearTo: 2015,
        },
        reason: "YEAR_NOT_IN_RANGE",
        siblingReports: [
          {
            model: "Accord VI, VII et VIII",
            yearFrom: 1998,
            yearTo: 2015,
          },
          {
            model: "Civic VI, VII, VIII et IX",
            yearFrom: 1996,
            yearTo: 2015,
          },
          {
            model: "CR-V I, II, III et IV",
            yearFrom: 1997,
            yearTo: 2015,
          },
          {
            model: "FR-V",
            yearFrom: 2004,
            yearTo: 2009,
          },
          {
            model: "Insight II",
            yearFrom: 2007,
            yearTo: 2014,
          },
          {
            model: "Legend IV",
            yearFrom: 2007,
            yearTo: 2010,
          },
          {
            model: "Stream",
            yearFrom: 2004,
            yearTo: 2005,
          },
        ],
        splitVIN: {
          modelYear: "H",
          vds: "GK3770",
          vis: "HX000000",
          wmi: "JHM",
        },
      });
    });
  });

  describe("manufacturer mapping", () => {
    let libEUManufacturers: string[] = [];

    beforeAll(() => {
      libEUManufacturers = buildLibEuropeanManufacturers();
    });

    it("should support all EU manufacturers", () => {
      const map: Record<string, Manufacturer | undefined> = {};
      for (const manufacturer of libEUManufacturers) {
        map[manufacturer] = mapLibManufacturer(manufacturer);
      }
      expect(map).toMatchSnapshot();
    });
  });
});
