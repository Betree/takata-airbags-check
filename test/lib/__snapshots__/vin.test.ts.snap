// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VIN checker manufacturer mapping should support all EU manufacturers 1`] = `
{
  "ARO": undefined,
  "ASA-LIFT: Vegetable Technology Specialist (Agricultural Machinery)": undefined,
  "AZLK": undefined,
  "Adam Opel AG/Vauxhall & Holden": "opel",
  "Alexander Dennis": undefined,
  "Alfa Romeo Alfasud & Sprint through 1989": undefined,
  "Alfa Romeo car & Nissan Cherry Europe": "nissan",
  "Alpina": undefined,
  "Alpine": undefined,
  "Aprilia": undefined,
  "Arbeitsgemeinschaft VW-MAN": undefined,
  "Aston Martin Lagonda Limited": undefined,
  "Aston Martin SUVs": undefined,
  "Astra Bus": undefined,
  "Audi": "audi",
  "Audi SUV": "audi",
  "Audi Sport GmbH SUV": "audi",
  "Audi car": "audi",
  "Audi/Porsche RS 2 Avant": "audi",
  "Austin": undefined,
  "Austin Rover Group including Sterling Cars": undefined,
  "Autobianchi": undefined,
  "Autobusfabriek Bova BV": undefined,
  "Automobili Pininfarina": undefined,
  "Avia/Daewoo Avia": undefined,
  "AvtoVAZ": undefined,
  "BMW M car": "bmw",
  "BMW Motorrad Motorcycle": "bmw",
  "BMW Motorrad Motorcycles made in India": "bmw",
  "BMW Motorrad Motorscooters made in China": "bmw",
  "BMW SUV": "bmw",
  "BMW car": "bmw",
  "BMW i SUV": "bmw",
  "BMW i car": "bmw",
  "BMW made by Avtotor": "bmw",
  "Bedford Vehicles": undefined,
  "Benelli": undefined,
  "Bentley Motors Limited": undefined,
  "Bentley SUV": undefined,
  "Betamotor S.p.A.": undefined,
  "Bimota": undefined,
  "Bitter Cars": undefined,
  "Bluecar": undefined,
  "Bogdan group": undefined,
  "Brabus": undefined,
  "Bremach": undefined,
  "Bucher Municipal including Johnston Sweepers of the UK (owned by Bucher Municipal of Switzerland)": undefined,
  "Bugatti Automobiles S.A.S.": undefined,
  "Bugatti Automobili S.p.A": undefined,
  "Buick Regal & Buick Cascada": undefined,
  "Cacciamali": undefined,
  "Cadillac BLS built by Saab": "cadillac",
  "Cadillac Catera": "cadillac",
  "Cagiva & MV Agusta": undefined,
  "Caterham Cars": undefined,
  "Chevrolet made by Avtotor": undefined,
  "Chrysler TC by Maserati": "chrysler",
  "Citro\\xEBn": undefined,
  "Cizeta Automobili SRL": undefined,
  "DAF Bus": undefined,
  "DAF Trucks & Leyland DAF": undefined,
  "DR Automobiles": undefined,
  "DS Automobiles": "stellantis",
  "Dacia": undefined,
  "Daewoo Romania": undefined,
  "Daimler until April 1987": undefined,
  "Dallara Stradale": undefined,
  "De Tomaso Modena SpA": undefined,
  "DeLorean": undefined,
  "Dennis": undefined,
  "Dennis Eagle": undefined,
  "Derbi": undefined,
  "Dodge Sprinter MPV (van with 2 or 3 rows of seats) 2005-2009": undefined,
  "Dodge Sprinter bus (van with more than 3 rows of seats) 2008-2009": undefined,
  "Dodge Sprinter incomplete vehicle 2005-2009": undefined,
  "Dodge Sprinter truck (cargo van with 1 row of seats) 2005-2009": undefined,
  "Dodge Trucks - UK 1981-1984": undefined,
  "Dr\\xF6gm\\xF6ller": undefined,
  "Ducati Motor Holding": undefined,
  "ERF": undefined,
  "Effedi Veicoli Commerciali": undefined,
  "Energica Motor Company": undefined,
  "Euro Bus Diamond": undefined,
  "EvoBus Mercedes-Benz buses": "mercedes",
  "FSC Lublin Automotive Factory": undefined,
  "FSD\\xA0[pl]": undefined,
  "Fabryka Samochod\\xF3w Ma\\u0142olitra\\u017Cowych": undefined,
  "Fabryka Samochodow Osobowych": undefined,
  "Fantic Motor": undefined,
  "Ferrari": "ferrari",
  "Fiat MPV/SUV": undefined,
  "Fiat car Also Fiat Fullback": undefined,
  "Fisker Automotive (Fisker Karma) built by Valmet Automotive": "fisker",
  "Fisker Inc. (Fisker Ocean) made by Magna Steyr": "fisker",
  "Foden": undefined,
  "Ford Espana": "ford",
  "Ford Motor Company ZAO": "ford",
  "Ford Sollers": "ford",
  "Ford of Britain": "ford",
  "Ford of Europe": "ford",
  "Freight Rover": undefined,
  "Freightliner Econic": undefined,
  "Freightliner Sprinter 2002 & Sprinter (Dodge or Freightliner) 2003-2005 MPV (van with 2 or 3 rows of seats)": undefined,
  "Freightliner Sprinter 2002 & Sprinter (Dodge or Freightliner) 2003-2005 incomplete vehicle": undefined,
  "Freightliner Sprinter 2002 & Sprinter (Dodge or Freightliner) 2003-2005 truck (cargo van with 1 row of seats)": undefined,
  "Freightliner Sprinter MPV (van with 2 or 3 rows of seats)": undefined,
  "Freightliner Sprinter MPV (van with 2 or 3 rows of seats) 2005-2019": undefined,
  "Freightliner Sprinter bus (van with more than 3 rows of seats)": undefined,
  "Freightliner Sprinter bus (van with more than 3 rows of seats) 2008-2019": undefined,
  "Freightliner Sprinter incomplete vehicle": undefined,
  "Freightliner Sprinter incomplete vehicle 2005-2019": undefined,
  "Freightliner Sprinter truck (cargo van with 1 row of seats)": undefined,
  "Freightliner Sprinter truck (cargo van with 1 row of seats) 2005-2019": undefined,
  "Freightliner Unimog incomplete vehicle": undefined,
  "Freightliner Unimog truck": undefined,
  "GAZ": undefined,
  "GM-AvtoVAZ": undefined,
  "Gas Gas": undefined,
  "General Motors De Portugal Lda": undefined,
  "General Motors Russia": undefined,
  "GolAZ": undefined,
  "Great Wall Motors IMS Avtotrade-12 LLC": undefined,
  "Great Wall Motors KrASZ": undefined,
  "Great Wall Motors/Litex Motors AD": undefined,
  "Gruppo Bertone": undefined,
  "HWA AG": undefined,
  "Heuliez Bus": undefined,
  "Honda": "honda",
  "Honda Italia Industriale S.p.A.": "honda",
  "Honda UK Manufacturing SUV": "honda",
  "Honda UK Manufacturing car": "honda",
  "Honda made by Austin Rover Group": "honda",
  "Husaberg Swedish built models": undefined,
  "Husqvarna Motorcycles Srl Under BMW ownership": "bmw",
  "Husqvarna Motorcycles Under Cagiva ownership": undefined,
  "Hyundai Automobili Italia Importazioni": undefined,
  "Hyundai Motor Company & Kia made by Avtotor": undefined,
  "Hyundai Motor Manufacturing Czech SUV": undefined,
  "Hyundai Motor Manufacturing Czech car": undefined,
  "Hyundai Motor Manufacturing Rus": undefined,
  "IBC Vehicles (Isuzu Bedford Company)": undefined,
  "IMZ-Ural Ural Motorcycles": undefined,
  "INEOS Automotive": undefined,
  "Ifor Williams Trailers": undefined,
  "Ikarus Bus": undefined,
  "Innocenti": undefined,
  "Irmscher": undefined,
  "Italdesign Automobili Speciali": undefined,
  "Italjet": undefined,
  "Iveco": undefined,
  "Iveco Bus": undefined,
  "Iveco Defence": undefined,
  "Iveco Ford Truck": "ford",
  "Iveco Unic": undefined,
  "Iveco/Iveco Magirus": undefined,
  "JC Bamford Excavators": undefined,
  "Jaguar": "jaguar",
  "Jawa Moto": undefined,
  "Jeep and Dodge Hornet": "jeep",
  "Jelcz": undefined,
  "Jonckheere": undefined,
  "KAMAZ": undefined,
  "KAvZ": undefined,
  "KTM": undefined,
  "Karosa": undefined,
  "Kenworth & Peterbilt trucks made by Leyland Trucks": undefined,
  "Kia Motors Slovakia SUV": undefined,
  "Kia Motors Slovakia car": undefined,
  "Koenigsegg Automotive AB": undefined,
  "LDV Group": undefined,
  "Lamborghini SUV": undefined,
  "Lamborghini car from mid 2003": undefined,
  "Lamborghini through mid 2003": undefined,
  "Lancia": undefined,
  "Land Rover": "land rover",
  "Laverda": undefined,
  "Leyland Bus": undefined,
  "Leyland Motors & Leyland DAF": undefined,
  "Leyland Vehicles": undefined,
  "LiAZ": undefined,
  "London Electric Vehicle Company & London Taxi Company & London Taxis International": undefined,
  "Lotus Cars Limited": undefined,
  "Lynx (snowmobile)": undefined,
  "MAN": undefined,
  "MBK & Yamaha Motor": undefined,
  "MG Motor NAC MG UK & MG Motor UK": undefined,
  "MG Sport and Racing": undefined,
  "MLT Automotive": undefined,
  "MZ Motorrad- und Zweiradwerk": undefined,
  "Malaguti": undefined,
  "Marcos Engineering": undefined,
  "Maserati SUV": undefined,
  "Maserati car": undefined,
  "Mathieu Fayat Group": undefined,
  "Matra/Talbot/Simca": undefined,
  "McLaren F1": undefined,
  "McLaren Group": undefined,
  "Mercedes-AMG used for Mercedes-Benz SLS AMG & Mercedes-AMG GT (not used in North America)": "mercedes",
  "Mercedes-Benz & Maybach": "mercedes",
  "Mercedes-Benz Espana SA (Vitoria)": "mercedes",
  "Mercedes-Benz MPV (van with 2 or 3 rows of seats) (North America)": "mercedes",
  "Mercedes-Benz SUV": "mercedes",
  "Mercedes-Benz Trucks Vostok": "mercedes",
  "Mercedes-Benz bus (van with more than 3 rows of seats) (North America)": "mercedes",
  "Mercedes-Benz car": "mercedes",
  "Mercedes-Benz incomplete vehicle (North America)": "mercedes",
  "Mercedes-Benz truck": "mercedes",
  "Mercedes-Benz truck (cargo van with 1 row of seats) (North America)": "mercedes",
  "Mercedes-Benz van": "mercedes",
  "Mercedes-Benz van (French & Spanish built models - Citan & Vito & X-Class)": "mercedes",
  "Merkur": undefined,
  "Metal-Fach Sp. z o.o.": undefined,
  "Micro Compact Car (until 5/99)": undefined,
  "Mini Moke made by British Leyland & Austin Rover": undefined,
  "Mini Moke made by Cagiva": undefined,
  "Mini SUV": undefined,
  "Mini car": undefined,
  "Mitsubishi Fuso Truck and Bus Corp.": "mitsubishi",
  "Montesa Honda Honda Montesa motorcycle models": "honda",
  "Montesa Honda Honda motorcycle models": "honda",
  "Morgan 3-Wheeler": undefined,
  "Morgan Motor Company": undefined,
  "Morris": undefined,
  "Moto Guzzi": undefined,
  "Moto Morini srl": undefined,
  "Mowag": undefined,
  "NedCar B.V. Mitsubishi Motors": "mitsubishi",
  "NedCar B.V. Mitsubishi Motors (Colt CZC convertible)": "mitsubishi",
  "Neoplan": undefined,
  "Next.e.GO Mobile SE": undefined,
  "Nissan": "nissan",
  "Nissan Manufacturing Rus": "nissan",
  "Nissan Motor Iberica SA, Nissan passenger car/MPV/van/SUV/pickup & Ford Maverick 1993-1999": "ford",
  "Nissan Motor Manufacturing UK Infiniti": "nissan",
  "Nissan Motor Manufacturing UK Nissan": "nissan",
  "Nissan Vehiculos Industriales SA, Nissan Commercial Vehicles": "nissan",
  "Norton Motorcycles": undefined,
  "Oltcit": undefined,
  "Opel Automobile Espana,SA": "opel",
  "Opel Automobile Gmbh/Vauxhall & Holden": "opel",
  "Opel Automobile Gmbh/Vauxhall car/SUV": "opel",
  "Opel Automobile Gmbh/Vauxhall van": "opel",
  "Opel Special Vehicles": "opel",
  "Optare": undefined,
  "PAZ": undefined,
  "PCMA Rus": undefined,
  "Pagani Automobili S.p.A.": undefined,
  "Panther": undefined,
  "Pegaso": undefined,
  "Peugeot": "peugeot",
  "Peugeot Motocycles": "peugeot",
  "Piaggio and Vespa and Gilera": undefined,
  "Pininfarina": undefined,
  "Polaris Europe (based in Switzerland)": undefined,
  "Polestar": undefined,
  "Porsche SUV": undefined,
  "Porsche car": undefined,
  "Qvale": undefined,
  "Ram 1200": undefined,
  "Rayton-Fissore Magnum": undefined,
  "Renault & Eagle Medallion made by Renault": "renault",
  "Renault Espa\\xF1a": "renault",
  "Renault Portuguesa SARL": "renault",
  "Renault Russia": "renault",
  "Renault SOVAB": "renault",
  "Renault Sport Spider": "renault",
  "Renault Trucks": "renault",
  "Renault Trucks & Mack Trucks medium duty trucks made by Renault Trucks": "renault",
  "Renault Trucks Industries 1985-1992": "renault",
  "Rimac Automobili": undefined,
  "Rocar": undefined,
  "Rolls-Royce Motor Cars & Rolls-Royce Motors car": undefined,
  "Rolls-Royce Motor Cars SUV": undefined,
  "Rover & MG Rover Group": undefined,
  "Ruf Automobile": undefined,
  "SEAT": "seat",
  "SOR Libchavy": undefined,
  "SWM Motorcycles S.r.l.": undefined,
  "Saab": undefined,
  "Saab NEVS": undefined,
  "Saab built by Valmet Automotive": undefined,
  "Salvador Caetano": undefined,
  "Sanocka Fabryka Autobusow Sfa / Autosan": undefined,
  "Santana Motor": undefined,
  "Saturn Astra": undefined,
  "Scammell": undefined,
  "Scania Nederland B.V.": undefined,
  "Scania Production Angers": undefined,
  "Scania Production Slupsk": undefined,
  "Scania, Katrineholm bus": undefined,
  "Scania, S\\xF6dert\\xE4lje truck": undefined,
  "Setra & Karl K\\xE4ssbohrer Fahrzeugwerke": undefined,
  "Sisu Auto": undefined,
  "Smart": undefined,
  "Smart (from 5/99)": undefined,
  "Solaris Bus & Coach": undefined,
  "Spyker Cars": undefined,
  "Steyr-Daimler-Puch Puch G & Steyr-Puch Pinzgauer": undefined,
  "Steyr-Daimler-Puch Steyr Trucks": undefined,
  "StreetScooter": undefined,
  "Suzuki": "suzuki",
  "Suzuki Motor Espa\\xF1a Motorcycle": "suzuki",
  "Swift Group Ltd.": undefined,
  "TVR": undefined,
  "Talbot": undefined,
  "Talbot Motor Company": undefined,
  "Tatra car": undefined,
  "Tatra trucks": undefined,
  "Tauro Sport Auto": undefined,
  "Tesla Europe (based in the Netherlands)": "tesla",
  "Tesla Roadster made by Lotus": "tesla",
  "Think Nordic": undefined,
  "Tomos motorcycle": undefined,
  "Toyota Caetano Portugal": "toyota",
  "Toyota Manufacturing UK": "toyota",
  "Toyota Motor Europe (based in Belgium) used for Toyota ProAce & Toyota ProAce City": "toyota",
  "Toyota Motor Manufacturing France": "toyota",
  "Toyota Motor Manufacturing Russia": "toyota",
  "Toyota Supra Fifth generation": "toyota",
  "Trabant": undefined,
  "Triumph Engineering original Triumph Motorcycle company": undefined,
  "Triumph Motorcycles Ltd current Triumph Motorcycle company": undefined,
  "Triumph car": undefined,
  "Tvornica Autobusa Zagreb (TAZ)": undefined,
  "UAZ": undefined,
  "UAZ Special Purpose Vehicles": undefined,
  "UTB (Uzina de Tractoare Bra\\u0219ov)": undefined,
  "UzDaewoo GM Uzbekistan Ravon": undefined,
  "VDL Bova": undefined,
  "VDL Bus & Coach": undefined,
  "VDL Bus International": undefined,
  "Van Hool": undefined,
  "Vauxhall Motors": undefined,
  "Venturi Automobiles": undefined,
  "Viseon Bus": undefined,
  "Volkswagen Commercial Vehicles": "volkswagen",
  "Volkswagen Group Rus": "volkswagen",
  "Volkswagen SUV": "volkswagen",
  "Volkswagen car": "volkswagen",
  "Volvo Bus": undefined,
  "Volvo Buses": undefined,
  "Volvo Car B.V./NedCar B.V. (Volvo Cars)": undefined,
  "Volvo Cars SUV": undefined,
  "Volvo Cars car": undefined,
  "Volvo Truck & Bus": undefined,
  "Volvo Trucks": undefined,
  "Volvo Trucks incomplete vehicle": undefined,
  "Volvo Vostok ZAO Volvo Trucks": undefined,
  "Wartburg": undefined,
  "Westfield Sportscars": undefined,
  "Wielton": undefined,
  "Yamaha Motor Espa\\xF1a SA & Yamaha Sociedad Espa\\xF1ola de Motocicletas": undefined,
  "Yamaha Motor Italia SpA & Belgarda SpA": undefined,
  "ZAZ": undefined,
  "Zastava Automobiles": undefined,
  "\\u0160koda Auto AS": undefined,
  "\\u0160koda Auto made by Eurocar": undefined,
  "\\u0160koda Ostrov": undefined,
  "\\xD6AF": undefined,
  "e.GO Mobile AG": undefined,
}
`;
